File: test/Models/pheno_pkg/src/openalea/Ptq.py
# coding: utf8
from copy import copy

import numpy
from math import *

def model_ptq(tTWindowForPTQ = 70.0,
         kl = 0.45,
         listTTShootWindowForPTQ_t1 = [0.0],
         listPARTTWindowForPTQ_t1 = [0.0],
         listGAITTWindowForPTQ = [0.0, 5.2],
         pAR = 0.0,
         deltaTT = 0.0):
    """
     - Name: PTQ -Version: 1.0, -Time step: 1
     - Description:
                 * Title: Phyllochron Model
                 * Author: Pierre Martre
                 * Reference: Modeling development phase in the 
                     Wheat Simulation Model SiriusQuality.
                     See documentation at http://www1.clermont.inra.fr/siriusquality/?page_id=427
                 * Institution: INRA Montpellier
                 * Abstract: Calculate Photothermal Quaotient 
     - inputs:
                 * name: tTWindowForPTQ
                               ** parametercategory : species
                               ** min : 0.0
                               ** datatype : DOUBLE
                               ** max : 70000.0
                               ** uri : some url
                               ** default : 70.0
                               ** inputtype : parameter
                               ** unit : °C d
                               ** description : Thermal time window in which averages are computed
                 * name: kl
                               ** parametercategory : species
                               ** min : 0.0
                               ** datatype : DOUBLE
                               ** max : 50.0
                               ** uri : some url
                               ** default : 0.45
                               ** inputtype : parameter
                               ** unit : 
                               ** description : Exctinction Coefficient
                 * name: listTTShootWindowForPTQ_t1
                               ** min : 
                               ** default : [0.0]
                               ** max : 
                               ** uri : some url
                               ** variablecategory : state
                               ** datatype : DOUBLELIST
                               ** inputtype : variable
                               ** unit : °C d
                               ** description : List of Daily Shoot thermal time during TTWindowForPTQ thermal time period
                 * name: listPARTTWindowForPTQ_t1
                               ** min : 
                               ** default : [0.0]
                               ** max : 
                               ** uri : some url
                               ** variablecategory : state
                               ** datatype : DOUBLELIST
                               ** inputtype : variable
                               ** unit : MJ m2 d
                               ** description : List of Daily PAR during TTWindowForPTQ thermal time period
                 * name: listGAITTWindowForPTQ
                               ** min : 
                               ** default : [0.0, 5.2]
                               ** max : 
                               ** uri : some url
                               ** variablecategory : state
                               ** datatype : DOUBLELIST
                               ** inputtype : variable
                               ** unit : m2 m-2
                               ** description : List of daily GAI over TTWindowForPTQ thermal time period
                 * name: pAR
                               ** min : 0.0
                               ** default : 0.0
                               ** max : 10000.0
                               ** uri : some url
                               ** variablecategory : auxiliary
                               ** datatype : DOUBLE
                               ** inputtype : variable
                               ** unit : MJ m² d
                               ** description : Daily Photosyntetically Active Radiation
                 * name: deltaTT
                               ** min : 0.0
                               ** default : 0.0
                               ** max : 100.0
                               ** uri : some url
                               ** variablecategory : auxiliary
                               ** datatype : DOUBLE
                               ** inputtype : variable
                               ** unit : °C d
                               ** description : daily delta TT 
     - outputs:
                 * name: listPARTTWindowForPTQ
                               ** datatype : DOUBLELIST
                               ** min : 
                               ** variablecategory : state
                               ** max : 
                               ** unit : MJ m2 d
                               ** description :  List of Daily PAR during TTWindowForPTQ thermal time period
                 * name: listTTShootWindowForPTQ
                               ** datatype : DOUBLELIST
                               ** min : 
                               ** variablecategory : state
                               ** max : 
                               ** unit : °C d
                               ** description : List of Daily Shoot thermal time during TTWindowForPTQ thermal time period
                 * name: ptq
                               ** datatype : DOUBLE
                               ** min : 0
                               ** variablecategory : state
                               ** max : 10000
                               ** unit : MJ °C-1 d m-2)
                               ** description : Photothermal Quotient
    """

    listPARTTWindowForPTQ = []
    listTTShootWindowForPTQ = []
    TTList = []
    PARList = []
    parInt = 0.0
    for i in range(0 , len(listTTShootWindowForPTQ_t1) , 1):
        TTList.append(listTTShootWindowForPTQ_t1[i])
        PARList.append(listPARTTWindowForPTQ_t1[i])
    TTList.append(deltaTT)
    PARList.append(pAR)
    SumTT = sum(TTList)
    count = 0
    while SumTT > tTWindowForPTQ:
        SumTT = SumTT - TTList[count]
        count = count + 1
    for i in range(count , len(TTList) , 1):
        listTTShootWindowForPTQ.append(TTList[i])
        listPARTTWindowForPTQ.append(PARList[i])
    for i in range(0 , len(listTTShootWindowForPTQ) , 1):
        parInt = parInt + (listPARTTWindowForPTQ[i] * (1 - exp(-kl * listGAITTWindowForPTQ[i])))
    TTShoot = sum(listTTShootWindowForPTQ)
    ptq = parInt / TTShoot
    return (listPARTTWindowForPTQ, listTTShootWindowForPTQ, ptq)

File: test/Models/pheno_pkg/src/f90/Ptq.f90
MODULE Ptqmod
    USE list_sub
    IMPLICIT NONE
CONTAINS

    SUBROUTINE model_ptq(tTWindowForPTQ, &
        kl, &
        listTTShootWindowForPTQ_t1, &
        listPARTTWindowForPTQ_t1, &
        listGAITTWindowForPTQ, &
        pAR, &
        deltaTT, &
        listPARTTWindowForPTQ, &
        listTTShootWindowForPTQ, &
        ptq)
        IMPLICIT NONE
        REAL, INTENT(IN) :: tTWindowForPTQ
        REAL, INTENT(IN) :: kl
        REAL, ALLOCATABLE , DIMENSION(:), INTENT(IN) ::  &
                listTTShootWindowForPTQ_t1
        REAL, ALLOCATABLE , DIMENSION(:), INTENT(IN) ::  &
                listPARTTWindowForPTQ_t1
        REAL, ALLOCATABLE , DIMENSION(:), INTENT(IN) :: listGAITTWindowForPTQ
        REAL, INTENT(IN) :: pAR
        REAL, INTENT(IN) :: deltaTT
        REAL, ALLOCATABLE , DIMENSION(:), INTENT(OUT) :: listPARTTWindowForPTQ
        REAL, ALLOCATABLE , DIMENSION(:), INTENT(OUT) ::  &
                listTTShootWindowForPTQ
        REAL, INTENT(OUT) :: ptq
        REAL, ALLOCATABLE , DIMENSION(:):: TTList
        REAL, ALLOCATABLE , DIMENSION(:):: PARList
        INTEGER:: i
        INTEGER:: count
        REAL:: SumTT
        REAL:: parInt
        REAL:: TTShoot
        parInt = 0.0
        !- Description:
    !            * Title: Phyllochron Model
    !            * Author: Pierre Martre
    !            * Reference: Modeling development phase in the 
    !                Wheat Simulation Model SiriusQuality.
    !                See documentation at http://www1.clermont.inra.fr/siriusquality/?page_id=427
    !            * Institution: INRA Montpellier
    !            * Abstract: Calculate Photothermal Quaotient 
        !- inputs:
    !            * name: tTWindowForPTQ
    !                          ** description : Thermal time window in which averages are computed
    !                          ** parametercategory : species
    !                          ** datatype : DOUBLE
    !                          ** min : 0.0
    !                          ** max : 70000.0
    !                          ** default : 70.0
    !                          ** unit : °C d
    !                          ** uri : some url
    !                          ** inputtype : parameter
    !            * name: kl
    !                          ** description : Exctinction Coefficient
    !                          ** parametercategory : species
    !                          ** datatype : DOUBLE
    !                          ** min : 0.0
    !                          ** max : 50.0
    !                          ** default : 0.45
    !                          ** unit : 
    !                          ** uri : some url
    !                          ** inputtype : parameter
    !            * name: listTTShootWindowForPTQ_t1
    !                          ** description : List of Daily Shoot thermal time during TTWindowForPTQ thermal time period
    !                          ** variablecategory : state
    !                          ** datatype : DOUBLELIST
    !                          ** min : 
    !                          ** max : 
    !                          ** default : [0.0]
    !                          ** unit : °C d
    !                          ** uri : some url
    !                          ** inputtype : variable
    !            * name: listPARTTWindowForPTQ_t1
    !                          ** description : List of Daily PAR during TTWindowForPTQ thermal time period
    !                          ** variablecategory : state
    !                          ** inputtype : variable
    !                          ** datatype : DOUBLELIST
    !                          ** min : 
    !                          ** max : 
    !                          ** default : [0.0]
    !                          ** unit : MJ m2 d
    !                          ** uri : some url
    !            * name: listGAITTWindowForPTQ
    !                          ** description : List of daily GAI over TTWindowForPTQ thermal time period
    !                          ** variablecategory : state
    !                          ** inputtype : variable
    !                          ** datatype : DOUBLELIST
    !                          ** min : 
    !                          ** max : 
    !                          ** default : [0.0, 5.2]
    !                          ** unit : m2 m-2
    !                          ** uri : some url
    !            * name: pAR
    !                          ** description : Daily Photosyntetically Active Radiation
    !                          ** variablecategory : auxiliary
    !                          ** datatype : DOUBLE
    !                          ** default : 0.0
    !                          ** min : 0.0
    !                          ** max : 10000.0
    !                          ** unit : MJ m² d
    !                          ** uri : some url
    !                          ** inputtype : variable
    !            * name: deltaTT
    !                          ** description : daily delta TT 
    !                          ** variablecategory : auxiliary
    !                          ** inputtype : variable
    !                          ** datatype : DOUBLE
    !                          ** min : 0.0
    !                          ** max : 100.0
    !                          ** default : 0.0
    !                          ** unit : °C d
    !                          ** uri : some url
        !- outputs:
    !            * name: listPARTTWindowForPTQ
    !                          ** description :  List of Daily PAR during TTWindowForPTQ thermal time period
    !                          ** variablecategory : state
    !                          ** datatype : DOUBLELIST
    !                          ** min : 
    !                          ** max : 
    !                          ** unit : MJ m2 d
    !            * name: listTTShootWindowForPTQ
    !                          ** description : List of Daily Shoot thermal time during TTWindowForPTQ thermal time period
    !                          ** variablecategory : state
    !                          ** datatype : DOUBLELIST
    !                          ** min : 
    !                          ** max : 
    !                          ** unit : °C d
    !            * name: ptq
    !                          ** description : Photothermal Quotient
    !                          ** variablecategory : state
    !                          ** datatype : DOUBLE
    !                          ** min : 0
    !                          ** max : 10000
    !                          ** unit : MJ °C-1 d m-2)
        DO i = 0 , SIZE(listTTShootWindowForPTQ_t1)-1, 1
            call Add(TTList, listTTShootWindowForPTQ_t1(i+1))
            call Add(PARList, listPARTTWindowForPTQ_t1(i+1))
        END DO
        call Add(TTList, deltaTT)
        call Add(PARList, pAR)
        SumTT = sum(TTList)
        count = 0
        DO WHILE ( SumTT .GT. tTWindowForPTQ )
            SumTT = SumTT - TTList(count+1)
            count = count + 1
        END DO
        DO i = count , SIZE(TTList)-1, 1
            call Add(listTTShootWindowForPTQ, TTList(i+1))
            call Add(listPARTTWindowForPTQ, PARList(i+1))
        END DO
        DO i = 0 , SIZE(listTTShootWindowForPTQ)-1, 1
            parInt = parInt + (listPARTTWindowForPTQ(i+1) * (1 - EXP((-kl) *  &
                    listGAITTWindowForPTQ(i+1))))
        END DO
        TTShoot = sum(listTTShootWindowForPTQ)
        ptq = parInt / TTShoot
    END SUBROUTINE model_ptq

END MODULE
